### DISCUSSION TITLE
Failed auto triggers for downstream jobs

### DISCUSSION THREAD

**Discussion by kirthiprakash** [concourse/concourse#7012] [Help & Support]
I had faced some issues with the auto triggering of downstream jobs (with `passed` and `version:every`). I have filed a bug report for the same here https://github.com/concourse/concourse/issues/6988.
The root cause for the issue is the presence of two consecutive builds on the upstream job with the same input versions.

Today I encountered the same issue but I couldn't find any builds on the upstream jobs with the same input versions and yet the scheduling algorithm was vouching for a candidate from an old build ID.
Upon further investigation, it was found that the old build ID belonged to a successful build just before a failed build. This is how the build history looked.

### Upstream Job
```
id        name                      status     start                     end                       duration  team  created by
9684253   qcs/stage-ml-engine/1213  succeeded  2021-05-12@01:54:58+0530  2021-05-12@02:14:21+0530  19m23s    main  system
9506814   qcs/stage-ml-engine/1212  succeeded  2021-05-11@21:08:51+0530  2021-05-11@21:22:59+0530  14m8s     main  system
9434680   qcs/stage-ml-engine/1211  succeeded  2021-05-11@12:27:31+0530  2021-05-11@12:41:59+0530  14m28s    main  system
9427130   qcs/stage-ml-engine/1210  succeeded  2021-05-11@11:35:04+0530  2021-05-11@11:46:18+0530  11m14s    main  system
9335512   qcs/stage-ml-engine/1209  succeeded  2021-05-11@00:54:03+0530  2021-05-11@01:09:08+0530  15m5s     main  system
9326813   qcs/stage-ml-engine/1208  succeeded  2021-05-10@23:51:49+0530  2021-05-11@00:05:00+0530  13m11s    main  system
9295007   qcs/stage-ml-engine/1207  succeeded  2021-05-10@20:15:37+0530  2021-05-10@20:30:44+0530  15m7s     main  system
9292830   qcs/stage-ml-engine/1206  succeeded  2021-05-10@20:04:30+0530  2021-05-10@20:15:33+0530  11m3s     main  system
9273491   qcs/stage-ml-engine/1205  failed     2021-05-10@17:43:56+0530  2021-05-10@18:10:20+0530  26m24s    main  system
9249018   qcs/stage-ml-engine/1204  succeeded  2021-05-10@14:40:53+0530  2021-05-10@15:00:05+0530  19m12s    main  system
9235551   qcs/stage-ml-engine/1203  succeeded  2021-05-10@13:08:50+0530  2021-05-10@13:24:30+0530  15m40s    main  system
9229745   qcs/stage-ml-engine/1202  succeeded  2021-05-10@12:27:35+0530  2021-05-10@12:40:29+0530  12m54s    main  system
8679138   qcs/stage-ml-engine/1201  succeeded  2021-05-07@20:13:55+0530  2021-05-07@20:28:35+0530  14m40s    main  system
```
### Downstream Job (`version:every` and `passed:stage-ml-engine`)
```
9843255   qcs/prod-ml-engine/946  succeeded  2021-05-12@20:40:49+0530  2021-05-12@20:42:10+0530  1m21s     main  system
9842855   qcs/prod-ml-engine/945  aborted    2021-05-12@20:38:34+0530  2021-05-12@20:39:45+0530  1m11s     main  user1
9817033   qcs/prod-ml-engine/944  succeeded  2021-05-12@17:33:11+0530  2021-05-12@17:34:31+0530  1m20s     main  system
9457515   qcs/prod-ml-engine/943  succeeded  2021-05-11@15:08:14+0530  2021-05-11@15:09:28+0530  1m14s     main  system
9457398   qcs/prod-ml-engine/942  succeeded  2021-05-11@15:04:47+0530  2021-05-11@15:06:00+0530  1m13s     main  user2
9455398   qcs/prod-ml-engine/941  succeeded  2021-05-11@14:50:34+0530  2021-05-11@14:52:11+0530  1m37s     main  user2
9251742   qcs/prod-ml-engine/940  succeeded  2021-05-10@15:00:09+0530  2021-05-10@15:01:25+0530  1m16s     main  system
9237941   qcs/prod-ml-engine/939  succeeded  2021-05-10@13:24:31+0530  2021-05-10@13:26:00+0530  1m29s     main  system
9231579   qcs/prod-ml-engine/938  succeeded  2021-05-10@12:40:33+0530  2021-05-10@12:41:54+0530  1m21s     main  system
```

Concourse is failing to trigger downstream job after the failed build at 9273491 (qcs/stage-ml-engine/120). For every tick of the scheduler, it is always picking the build 9249018 (one before the failed build) to resolve inputs for the downstream job.

<img width="1832" alt="Screenshot 2021-05-14 at 5 00 30 AM" src="https://user-images.githubusercontent.com/4361998/118200786-9640cd00-b473-11eb-9ec5-bf02b78c00c2.png">

The GroupResolver is trying to pick the build from the `build_pipes` table and this is what I found in the DB.

```
concourse=> select * from build_pipes where from_build_id=9249018;
 from_build_id | to_build_id 
---------------+-------------
       9249018 |     9251742
       9249018 |     9455398
       9249018 |     9457398
       9249018 |     9457515
       9249018 |     9817033
       9249018 |     9842855
       9249018 |     9843255
(8 rows)
```

Since it is not auto triggering, we are currently pinning the latest input version to trigger a build on the downstream job and that is what we see in the `build_pipes` table. I had not seen this behavior before and I feel something is wrong here. 

Do we know how can the scheduling algorithm reach such a state? 
Is there a way to recover? Can I safely remove some old build IDs from the `build_pipes`?
This issue is happening on production env and I was not able to reproduce this locally and hence little skeptical to modify the database tables.

**Comment by kirederik** [comment]
We are also seeing a similar issue with builds not triggering with new versions of the resource (specifically the telia-oss/git-pr-resource), unless we pin the version -- which automatically triggers the pending jobs.

**Reply by taylorsilva** [reply]
Are you seeing this with the most recent release, 7.4.0? The issue OP linked was fixed in that version.

**Reply by kirederik** [reply]
@taylorsilva yes. It may be a different issue but the symptoms are the same: build does not trigger unless the version is pinned. [This is our pipeline yaml](https://github.com/cloudfoundry-incubator/backup-and-restore-sdk-release/blob/a79c9a267aa4ae4756bf729c1e1b007512cf7998/ci/pipelines/backup-and-restore-sdk-release/pipeline.yml#L271-L274). The resulting can be seen [here](https://ci.cryo.cf-app.com/teams/bosh-backup-restore/pipelines/backup-and-restore-sdk-release) -- you can see how the latest version of [this resource](https://ci.cryo.cf-app.com/teams/bosh-backup-restore/pipelines/backup-and-restore-sdk-release/resources/backup-and-restore-sdk-release) failed to trigger the `merge-pr` job

**Reply by taylorsilva** [reply]
@kirederik any chance y'all can open an issue for this? I'm afraid it'll get lost in this discussion. As an issue it'll have a higher chance of being fixed by us or someone in the community.
